source/common/upstream/upstream_impl.cc

```diff
@@ -900,9 +900,25 @@ ClusterImplBase::ClusterImplBase(
   auto socket_factory = createTransportSocketFactory(cluster, factory_context);
   auto socket_matcher = std::make_unique<TransportSocketMatcherImpl>(
       cluster.transport_socket_matches(), factory_context, socket_factory, *stats_scope);
-  info_ = std::make_unique<ClusterInfoImpl>(cluster, factory_context.clusterManager().bindConfig(),
-                                            runtime, std::move(socket_matcher),
-                                            std::move(stats_scope), added_via_api, factory_context);
+  auto& dispatcher = factory_context.dispatcher();
+  info_ = std::shared_ptr<const ClusterInfoImpl>(
+      new ClusterInfoImpl(cluster, factory_context.clusterManager().bindConfig(), runtime,
+                          std::move(socket_matcher), std::move(stats_scope), added_via_api,
+                          factory_context),
+      [&dispatcher](const ClusterInfoImpl* self) {
+        FANCY_LOG(debug, "lambdai: schedule destroy cluster info {} on this thread", self->name());
+        if (!dispatcher.tryPost([self]() {
+          // TODO(lambdai): Yet there is risk that master dispatcher receives the function but doesn't execute during the shutdown.
+          // We can either 
+          // 1) Introduce folly::function which supports with unique_ptr capture and destroy cluster info by RAII, or
+          // 2) Call run post callback in master thread after no worker post back.
+          FANCY_LOG(debug, "lambdai: execute destroy cluster info {} on this thread. Master thread is expected.", self->name());
+          delete self;
+        })) {
+          FANCY_LOG(debug, "lambdai: cannot post. Has the master thread exited? Executing destroy cluster info {} on this thread.", self->name());
+          delete self;
```

Comment:
How can this happen? All workers should shut down and join before the main thread finishes running. Even if things are cleaned up after the join, it should be possibly to delete everything on the main thread. Perhaps in this case there needs to be some other cleanup/execution queue for posts that should be run even after the main thread dispatcher has exited?

Reply:
Yes, possible, I just want to raise the concern here we need some extra shutdown steps.

Reply:
Update: The current order is
1. master stop dispatching
2. shutdown TLS
3. stop workers

Master must refuse clusterInfo destroy closure before TLS shutdown(step2）. Otherwise the clusterInfo may trigger other TLS op and break TLS.

Alternatively we run some clean up in master queue, and disable TLS during the clean up.